File: demo-app/app/src/main/java/com/launchkey/android/authenticator/demo/ui/fragment/CustomDevicesFragment3.java
package com.launchkey.android.authenticator.demo.ui.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;

import com.launchkey.android.authenticator.sdk.SimpleOperationCallback;
import com.launchkey.android.authenticator.sdk.device.Device;
import com.launchkey.android.authenticator.sdk.device.DeviceManager;
import com.launchkey.android.authenticator.sdk.device.event.GetDevicesEventCallback;
import com.launchkey.android.authenticator.sdk.error.BaseError;
import com.launchkey.android.authenticator.demo.R;
import com.launchkey.android.authenticator.demo.ui.adapter.DemoDevicesAdapter;
import com.launchkey.android.authenticator.demo.util.Utils;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Created by armando on 7/8/16.
 */
public class CustomDevicesFragment3 extends BaseDemoFragment
        implements AdapterView.OnItemClickListener {

    private List<Device> mDevices = new ArrayList<>();
    private ListView mList;
    private DemoDevicesAdapter mAdapter;
    private DeviceManager mDeviceManager;
    private GetDevicesEventCallback mGetDevicesCallback;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View root = inflater.inflate(R.layout.demo_fragment_devices, container, false);
        return postInflationSetup(root);
    }

    private View postInflationSetup(View root) {
        mAdapter = new DemoDevicesAdapter(getActivity(), mDevices, this);

        mList = (ListView) root.findViewById(R.id.demo_fragment_devices_list);
        mList.setAdapter(mAdapter);

        return root;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mDeviceManager = DeviceManager.getInstance(getActivity());
        mGetDevicesCallback = new GetDevicesEventCallback() {

            @Override
            public void onEventResult(boolean successful, BaseError error, List<Device> devices) {

                if (successful) {
                    mDevices.clear();
                    mDevices.addAll(devices);
                    mAdapter.notifyDataSetChanged();
                } else {
                    Utils.simpleSnackbarForBaseError(mList, error);
                }
            }
        };
    }

    @Override
    public void onResume() {
        super.onResume();
        mDeviceManager.registerForEvents(mGetDevicesCallback);
        mDeviceManager.getDevices();
    }

    @Override
    public void onPause() {
        super.onPause();
        mDeviceManager.unregisterForEvents(mGetDevicesCallback);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {

        Device deviceToUnlink = mAdapter.getItem(position);
        mDeviceManager.unlinkDevice(deviceToUnlink, new SimpleOperationCallback<Device>() {

            @Override
            public void onResult(boolean successful, BaseError error, Device device) {

                if (successful) {
                    String message = String.format(Locale.getDefault(), "Device \"%s\" unlinked", device.getName());
                    Snackbar.make(mList, message, Snackbar.LENGTH_INDEFINITE)
                            .setAction("Return", new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    Utils.finish(CustomDevicesFragment3.this);
                                }
                            })
                            .show();
                    mDeviceManager.getDevices();
                } else {
                    Utils.simpleSnackbarForBaseError(mList, error);
                }
            }
        });
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/fragment/CustomDevicesFragment3.kt
package com.launchkey.android.authenticator.demo.ui.fragment

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ListView
import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.demo.ui.adapter.DemoDevicesAdapter
import com.launchkey.android.authenticator.demo.util.Utils
import com.launchkey.android.authenticator.sdk.device.Device
import com.launchkey.android.authenticator.sdk.device.DeviceManager
import com.launchkey.android.authenticator.sdk.device.event.GetDevicesEventCallback
import com.launchkey.android.authenticator.sdk.error.BaseError
import java.util.*

/**
 * Created by armando on 7/8/16.
 */
class CustomDevicesFragment3 : BaseDemoFragment(), AdapterView.OnItemClickListener {

    private val mDevices = ArrayList<Device>()
    private var mList: ListView? = null
    private var mAdapter: DemoDevicesAdapter? = null
    private var mDeviceManager: DeviceManager? = null
    private var mGetDevicesCallback: GetDevicesEventCallback? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val root = inflater.inflate(R.layout.demo_fragment_devices, container, false)
        return postInflationSetup(root)
    }

    private fun postInflationSetup(root: View): View {
        mAdapter = DemoDevicesAdapter(activity!!, mDevices, this)

        mList = root.findViewById<View>(R.id.demo_fragment_devices_list) as ListView
        mList!!.adapter = mAdapter

        return root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mDeviceManager = DeviceManager.getInstance(activity)
        mGetDevicesCallback = object : GetDevicesEventCallback() {

            override fun onEventResult(successful: Boolean, error: BaseError?, devices: List<Device>?) {

                if (successful) {
                    mDevices.clear()
                    mDevices.addAll(devices!!)
                    mAdapter!!.notifyDataSetChanged()
                } else {
                    Utils.simpleSnackbarForBaseError(mList, error)
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        mDeviceManager!!.registerForEvents<GetDevicesEventCallback>(mGetDevicesCallback)
        mDeviceManager!!.getDevices()
    }

    override fun onPause() {
        super.onPause()
        mDeviceManager!!.unregisterForEvents<GetDevicesEventCallback>(mGetDevicesCallback)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {

        val deviceToUnlink = mAdapter!!.getItem(position)
        mDeviceManager!!.unlinkDevice(deviceToUnlink) { successful, error, device ->
            if (successful) {
                val message = String.format(Locale.getDefault(), "Device \"%s\" unlinked", device.name)
                Snackbar.make(mList!!, message, Snackbar.LENGTH_INDEFINITE)
                        .setAction("Return") { Utils.finish(this@CustomDevicesFragment3) }
                        .show()
                mDeviceManager!!.getDevices()
            } else {
                Utils.simpleSnackbarForBaseError(mList, error)
            }
        }
    }
}
